### PR TITLE
Check invalid operand value

### PR DESCRIPTION
**Pull request by sunlin-link**

### CODE REVIEW

File: services/ml/model_impl_mac.mm [hunk 1]
```diff
@@ -107,7 +107,7 @@
     int32_t result = SetOperandValue(value_info->index,
                                      static_cast<const void*>(memory_.get() + value_info->offset),
                                      value_info->length);
-    if (result != mojom::NO_ERROR) {
+    if (result != mojom::NOT_ERROR) {
```
Comment (fujunwei) [inline services/ml/model_impl_mac.mm]: Maybe you need update source code, that has been fixed.

File: third_party/blink/renderer/modules/ml/model.cpp [hunk 2]
```diff
@@ -34,7 +34,37 @@ void Model::addOperand(const OperandOptions& options, ExceptionState& exception_
...
     return;
   }
+
   int32_t type = options.type();
+  switch (type) {
+    case NeuralNetworkContext::kFloat32:
+    case NeuralNetworkContext::kTensorFloat32:
+      if (!options.hasDimensions()) {
+        exception_state.ThrowDOMException(DOMExceptionCode::kInvalidStateError,
+            "Operand is invalid.");
```
Comment (fujunwei) [inline third_party/blink/renderer/modules/ml/model.cpp]: Maybe need use "git cl format"

File: third_party/blink/renderer/modules/ml/model.cpp [hunk 3]
```diff
@@ -34,7 +34,37 @@ void Model::addOperand(const OperandOptions& options, ExceptionState& exception_
...
+        return;
+      }
+      break;
+    case NeuralNetworkContext::kUint32:
+      break;
+    case NeuralNetworkContext::kTensorQuant8Asymm:
+      if (!options.hasDimensions() || !options.hasScale() || !options.hasZeroPoint()) {
```
Comment (fujunwei) [inline third_party/blink/renderer/modules/ml/model.cpp]: Use a flag to check the invalid state such, and the throw exception will at the end of the block such as 
 if (invalid_state) {
  exception_state.ThrowDOMException(DOMExceptionCode::kInvalidStateError,
                                                               Operand is invalid.");
  return;
}

File: third_party/blink/renderer/modules/ml/model.cpp [hunk 4]
```diff
@@ -34,7 +34,37 @@ void Model::addOperand(const OperandOptions& options, ExceptionState& exception_
...
+        exception_state.ThrowDOMException(DOMExceptionCode::kInvalidStateError,
+            "Operand is invalid.");
+        return;
+      }
+      break;
+    default:
+      break;
```
Comment (fujunwei) [inline third_party/blink/renderer/modules/ml/model.cpp]: If the default case should never execute, treat this as an error with assert(false);

File: third_party/blink/renderer/modules/ml/model.cpp [hunk 5]
```diff
@@ -75,34 +105,52 @@ void Model::setOperandValue(uint32_t index,
...
+    case WTF::ArrayBufferView::kTypeFloat32:
+      if (operand->type != NeuralNetworkContext::kFloat32 &&
+          operand->type != NeuralNetworkContext::kTensorFloat32) {
+        exception_state.ThrowDOMException(
+            DOMExceptionCode::kInvalidStateError, "Data type is invalid.");
+        return;
+      }
+      break;
+    default:
+      exception_state.ThrowDOMException(
```
Comment (fujunwei) [inline third_party/blink/renderer/modules/ml/model.cpp]: The default is unnecessary for enum type, because all of case has been handled.

**Comment by huningxin** [comment]
@sunlin-link , you may need to address review comments and rebase to webml branch ToT.

File: third_party/blink/renderer/modules/ml/model.cpp (hunk 5 above)
Comment (sunlin-link) [reply third_party/blink/renderer/modules/ml/model.cpp]: If someone add any new type to this enum and forget to check it here, the `default` case would be useful

File: third_party/blink/renderer/modules/ml/model.cpp [hunk 6]
```diff
@@ -70,6 +70,31 @@ void Model::addOperand(const OperandOptions& options,
...
+      if (!options.hasDimensions()) {
+        exception_state.ThrowDOMException(
+            DOMExceptionCode::kInvalidStateError, "Data type is invalid.");
+        return;
+      }
+      break;
+    case NeuralNetworkContext::kTensorQuant8Asymm:
+      if (!options.hasDimensions() || !options.hasScale() || !options.hasZeroPoint()
+          || options.scale < 0 || options.zeroPoint < 0 || options.zeroPoint > 255) {
+        exception_state.ThrowDOMException(
```
Comment (fujunwei) [inline third_party/blink/renderer/modules/ml/model.cpp]: How about using a flag like `bool invalid_date = false;` in order to reduce codes. And it's better to make message `Data type is invalid.` to a const like https://github.com/otcshare/chromium-src/blob/44b15cfd2af37d3695d055fd71125db3ffa3b376/third_party/blink/renderer/modules/ml/model.cpp#L22 , so l.124 can reuse the message.

if (invalid_date) {
    exception_state.ThrowDOMException(
            DOMExceptionCode::kInvalidStateError, "Data type is invalid.");
    return;
}

File: third_party/blink/renderer/modules/ml/model.cpp [hunk 7]
```diff
@@ -89,36 +114,41 @@ void Model::setOperandValue(uint32_t index,
...
-      (operand->type != NeuralNetworkContext::kTensorQuant8Asymm)) {
-    exception_state.ThrowDOMException(DOMExceptionCode::kInvalidStateError,
-                                      "Data type is invalid.");
-    return;
+  switch (options.type) {
+    case NeuralNetworkContext::kFloat32:
+    case NeuralNetworkContext::kTensorFloat32:
+      if (view_type != WTF::ArrayBufferView::kTypeFloat32) {
+        exception_state.ThrowDOMException(
+            DOMExceptionCode::kInvalidStateError, "Data type is invalid.");
```
Comment (fujunwei) [inline third_party/blink/renderer/modules/ml/model.cpp]: Ditto.

File: third_party/blink/renderer/modules/ml/model.cpp (hunk 6 above)
Comment (sunlin-link) [reply third_party/blink/renderer/modules/ml/model.cpp]: Once set invalid_state to true, no direct way to skip current checking and  to jump to the code that throws the exception. I will use InvalidState() to make the code more clean.

Comment (fujunwei) [reply third_party/blink/renderer/modules/ml/model.cpp]: Do you mean the "break" can't skip switch case?

**Comment by sunlin-link** [comment]
@fujunwei Updated

File: third_party/blink/renderer/modules/ml/model.h [hunk 8]
```diff
@@ -47,6 +47,9 @@ class Model final : public ScriptWrappable {
                            ml::mojom::blink::CompilationInitParamsPtr);
   void OnConnectionError();
 
+  bool checkOperand(const OperandOptions& options) const;
```
Comment (fujunwei) [inline third_party/blink/renderer/modules/ml/model.h]: Move to unnamespace. https://github.com/otcshare/chromium-src/blob/c484c416b48cfda9c85b1770884abd58c62a8519/third_party/blink/renderer/modules/ml/model.cpp#L20

File: third_party/blink/renderer/modules/ml/model.h [hunk 9]
```diff
@@ -47,6 +47,9 @@ class Model final : public ScriptWrappable {
...
+  bool checkOperandValue(const ml::mojom::blink::OperandPtr& operand, MaybeShared<DOMArrayBufferView> data) const;
```
Comment (fujunwei) [inline third_party/blink/renderer/modules/ml/model.h]: `const MaybeShared<DOMArrayBufferView>&`

File: third_party/blink/renderer/modules/ml/model.cpp [hunk 10]
```diff
@@ -69,6 +69,11 @@ void Model::addOperand(const OperandOptions& options,
                        : !options.hasType() ? "Operand type is missing." : "",
                    exception_state))
     return;
+ 
+  if (!checkOperand(options)) {
```
Comment (fujunwei) [inline third_party/blink/renderer/modules/ml/model.cpp]: I think it should be merged with l.67.

File: third_party/blink/renderer/modules/ml/model.cpp [hunk 11]
```diff
@@ -89,35 +94,8 @@ void Model::setOperandValue(uint32_t index,
...
-    exception_state.ThrowDOMException(DOMExceptionCode::kInvalidStateError,
-                                      "Data type is invalid.");
-    return;
-  }
-
-  if (view_type == WTF::ArrayBufferView::kTypeUint8 &&
-      (operand->type != NeuralNetworkContext::kTensorQuant8Asymm)) {
-    exception_state.ThrowDOMException(DOMExceptionCode::kInvalidStateError,
-                                      "Data type is invalid.");
+  if (!checkOperandValue(operand, data)) {
```
Comment (fujunwei) [inline third_party/blink/renderer/modules/ml/model.cpp]: It should be merged with l.88

File: third_party/blink/renderer/modules/ml/model.cpp (hunk 10 above)
Comment (fujunwei) [reply third_party/blink/renderer/modules/ml/model.cpp]: Do you have any update about this comment?

File: third_party/blink/renderer/modules/ml/model.cpp (hunk 11 above)
Comment (fujunwei) [reply third_party/blink/renderer/modules/ml/model.cpp]: Ditto @sunlin-link

File: third_party/blink/renderer/modules/ml/model.cpp [hunk 12]
```diff
@@ -19,36 +19,122 @@ namespace blink {
 
 namespace {
 
-constexpr char kModelFinishedError[] = "Model is finished.";
-
-bool InvalidState(const String& message, ExceptionState& exception_state) {
-  if (message.IsEmpty())
-    return false;
-
+void InvalidState(const String& message, ExceptionState& exception_state) {
```
Comment (fujunwei) [inline third_party/blink/renderer/modules/ml/model.cpp]: I think the code can readable if return true.

File: third_party/blink/renderer/modules/ml/model.cpp [hunk 13]
```diff
@@ -19,36 +19,122 @@ namespace blink {
...
   exception_state.ThrowDOMException(DOMExceptionCode::kInvalidStateError,
                                     message);
+}
+
+bool checkState(bool is_finished, ExceptionState& exception_state) {
```
Comment (fujunwei) [inline third_party/blink/renderer/modules/ml/model.cpp]: bool invalidState(bool is_finished, ExceptionState& exception_state) {
  return is_finished ? InvalidState("Model is finished", exception_state) : false;
}

File: third_party/blink/renderer/modules/ml/model.cpp [hunk 14]
```diff
@@ -19,36 +19,122 @@ namespace blink {
...
-
+bool checkInputOutput(size_t operands,
+                      const Vector<uint32_t>& inputs,
+                      const Vector<uint32_t>& outputs,
+                      ExceptionState& exception_state) {
   for (size_t i = 0; i < inputs.size(); ++i) {
-    if (inputs[i] > operands)
-      return InvalidState("Inputs is invalid.", exception_state);
+    if (inputs[i] >= operands) {
+      InvalidState("Inputs is invalid.", exception_state);
```
Comment (fujunwei) [inline third_party/blink/renderer/modules/ml/model.cpp]: bool invalidInputOutput

File: third_party/blink/renderer/modules/ml/model.cpp [hunk 15]
```diff
@@ -19,36 +19,122 @@ namespace blink {
...
+      return false;
+    }
   }
 
   for (size_t i = 0; i < outputs.size(); ++i) {
-    if (outputs[i] > operands)
-      return InvalidState("Outputs is invalid.", exception_state);
+    if (outputs[i] >= operands) {
+      InvalidState("Outputs is invalid.", exception_state);
```
Comment (fujunwei) [inline third_party/blink/renderer/modules/ml/model.cpp]: return InvalidState("Outputs is invalid.", exception_state);

File: third_party/blink/renderer/modules/ml/model.cpp [hunk 16]
```diff
@@ -19,38 +19,101 @@ namespace blink {
...
-                       const Vector<uint32_t>& inputs,
-                       const Vector<uint32_t>& outputs,
-                       ExceptionState& exception_state) {
-  if (operands == 0)
-    return InvalidState(kModelFinishedError, exception_state);
+bool InvalidAction(bool is_finished, ExceptionState& exception_state) {
+  if (is_finished)
+    return InvalidState("Model is finished", exception_state);
 
+  return false;
```
Comment (fujunwei) [inline third_party/blink/renderer/modules/ml/model.cpp]: return is_finished ? InvalidState("Model is finished", exception_state) : false

**Comment by fujunwei** [comment]
Thanks, LGTM.

**Pull request merged**